engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java

```diff
@@ -53,49 +47,39 @@ public ExecutionEnvironment withContextDisabled(Atom context) {
   }
 
   private ExecutionEnvironment update(Atom context, boolean value) {
-    assert context.getConstructor().getType()
-        == EnsoContext.get(null).getBuiltins().context().getType();
-    int keyFound = -1;
-    for (int i = 0; i < keys.length; i++) {
-      if (keys[i].equals(context.getConstructor().getName())) {
-        keyFound = i;
-      }
-    }
-    String[] keys1;
-    Boolean[] permissions1;
-    if (keyFound != -1) {
-      keys1 = cloneArray(keys, new String[keys.length]);
-      permissions1 = cloneArray(permissions, new Boolean[keys.length]);
-      permissions1[keyFound] = value;
+    var contextBuiltin = EnsoContext.get(null).getBuiltins().context();
+    assert context.getConstructor().getType() == contextBuiltin.getType();
+    var ctor = context.getConstructor();
+    Permissions newPermissions;
+    if (ctor == contextBuiltin.getInput()) {
```

Comment:
The use of `Permissions` seems promising. Consider _moving this code_ into the `Permissions` class itself.

Reply:
Refactored in dd25a6f256918a675898e23ca95172ab3d5d463a